Developer UX: Use <DateRange> component for Web server logs

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/1747

## Proposed Changes

Right now, in the Web server logs card, we use the following UI:

![server_logs_old](https://user-images.githubusercontent.com/497103/221234585-b42c481c-95bc-4cc7-bb84-407f95ae41e4.png)


This pr changes that, and lets the user select the dates that he wants from our already existing `<DateRange>` component:

![server_logs_new1](https://user-images.githubusercontent.com/497103/221236193-c9e2ee82-cdbe-470c-9a21-4f4083c223cb.png)

The original date validation is kept, as shown below:

![server_logs_new2](https://user-images.githubusercontent.com/497103/221236212-0f971c79-8249-4001-8cf2-54c065141401.png)

## Testing Instructions

### Log consistency

1. Create an atomic site.
2. Make sure that `ATOMIC SITE LOGS` is enabled in your user report card.
3. Checkout this pr and visit `Settings > Hosting Configuration` on your atomic site.
4. Do the same from `wordpress.com`
5. Select a date range
6. Download logs and compare the two.


### Date validation

1. Create an atomic site.
2. Make sure that `ATOMIC SITE LOGS` is enabled in your user report card.
3. Checkout this pr and visit `Settings > Hosting Configuration` on your atomic site.
4. Open the date range and try to create the following 3 scenarios:

**Enter an invalid date:**

- You should see something like: `End date format is not valid.` 
- Apply button doesn't update the dates.

**Enter a start date that is in the future in comparison with the end date:**

- You should see the following message: `Start date must be earlier than end date.`
- Apply button doesn't update the dates.

**Enter a start date that is longer than 14 days in the past from the end date:**

- You should see the following message: `Start date must be less than 14 days ago.`
- Apply button doesn't update the dates.

In all of the above cases, the `Download` button must be **disabled**.

### A small bug I found

1. Create an atomic site.
2. Make sure that `ATOMIC SITE LOGS` is enabled in your user report card.
3. Checkout this pr and visit `Settings > Hosting Configuration` on your atomic site.
4. Add the following code above the current 
5. Open the date range
6. Don't touch any input fields. Press the cancel button.
7. Dates should now be cleared.
8. Open the dev console.
9. You should not see any errors.

Remove this [function](https://github.com/Automattic/wp-calypso/pull/73774/files#diff-b6413809157f04a9569a6fb45bb11e80f0bda6beac8667edabdabce01d91f2d8R385)

and follow the exact steps above. 
You should see a console error.

## NOTES

Although the pr fixes the above issue, I've found two reasons that we may want to rethink that:

1. The dropdown date range picker is not that friendly for mobile users.
2. While changing the dates, I notice that there were times when the popover was moved, while I was editing the dates. This is not a good UX, and changing the behavior of the popover for example, is not something that can be taken lightly. This popover is used in other places, so we may break things in ways that cannot be easily identified.

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
sdfsdf
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

kozer committed:
Developer UX: Use <DateRange> component for Web server logs

sejas reviewed (changes requested):
I've found an issue using localized dates 
In Spanish we use : `DD/MM/YY` which raises the formatting error.

Steps to reproduce:
1. Change the UI language to Spanish on WPcom https://wordpress.com/me/account
2. Go to Hosting Configuration in an atomic site with a user with the RC logs activated
3. Open the datepicker
4. Change the date
5. Click on the input
6. Observe the error


https://user-images.githubusercontent.com/779993/221291401-8e6b007f-478a-420d-b0ba-c1acde9ff414.mp4

danielbachhuber commented on client/my-sites/hosting/web-server-logs-card/index.js:
```diff
@@ -18,6 +20,189 @@ import { getSelectedSiteId, getSelectedSiteSlug } from 'calypso/state/ui/selecto
+
+	return (
+		<FormFieldset className="date-range__date-inputs">
+			<legend className="date-range__date-inputs-legend">Start and End Dates</legend>
```
Does this string need to be translated?

danielbachhuber commented on client/my-sites/hosting/web-server-logs-card/index.js:
```diff
@@ -18,6 +20,189 @@ import { getSelectedSiteId, getSelectedSiteSlug } from 'calypso/state/ui/selecto
+						isError={ ! startValidationInfo.isValid }
+						onChange={ handleInputChange( 'Start' ) }
+						onBlur={ handleInputBlur( 'Start' ) }
+						onFocus={ handleInputFocus( 'Start' ) }
```
'Start' looks like a string that needs to be translated. If it's simply an argument, let's use `'start'`.

kozer replied:
Oh, that's interesting! Thanks for noticing.
This code is already inside the default `DateRangeInputs` component! So it's something that affects every page that we use the DateRange component! I'll fix that in there as well!

kozer replied:
This is part of the code, borrowed from `client/components/date-range/inputs.tsx`. It's used internally from the date-range component, and from the inputs component as well.

Do you think it's ok to refactor it and replace `Start` and `End` with `start` and `end`?

Update:
It's used to update textInputStartDate and textInputEndDate state variables. it doesn't make sense to refactor it. Eg:

```
		this.setState( {
			[ `textInput${ startOrEnd }Date` ]: val,
		} );

```

danielbachhuber replied:
> Do you think it's ok to refactor it and replace `Start` and `End` with `start` and `end`?
> 
> Update: It's used to update textInputStartDate and textInputEndDate state variables. it doesn't make sense to refactor it. Eg:

@kozer Ah, good catch. I agree, it doesn't make sense to refactor.

danielbachhuber reviewed (commented):
Interestingly, I'm able to select a date range longer than 14 days and downloading works fine:

<img width="1712" alt="image" src="https://user-images.githubusercontent.com/36432/221701186-ac78f854-0352-4940-ae4c-dcdb626da373.png">

Maybe we don't need the restriction?

I think it would be nice if the default was one week:

<img width="909" alt="image" src="https://user-images.githubusercontent.com/36432/221701423-7aa2ba6e-ee9d-4cb1-8768-75c5f9e297ab.png">

It also appears to be showing the date as UTC. I'd expect the date to appear as my timezone and then convert to UTC.

kozer committed:
Merge branch 'trunk' into fix/developer_ux_daterange_server_logs

kozer committed:
Fix: fix issues on different locales

kozer commented:
> I've found an issue using localized dates In Spanish we use : `DD/MM/YY` which raises the formatting error.
> 
> Steps to reproduce:
> 
> 1. Change the UI language to Spanish on WPcom https://wordpress.com/me/account
> 2. Go to Hosting Configuration in an atomic site with a user with the RC logs activated
> 3. Open the datepicker
> 4. Change the date
> 5. Click on the input
> 6. Observe the error

Can you please recheck this?

@danielbachhuber 

I refactored the code to disallow selecting dates 14 years ago on the calendar and inputs.
I have also updated the initial dates to be 1-week timeframe.

danielbachhuber commented on client/my-sites/hosting/web-server-logs-card/index.js:
```diff
@@ -18,6 +20,192 @@ import { getSelectedSiteId, getSelectedSiteSlug } from 'calypso/state/ui/selecto
+	return (
+		<FormFieldset className="date-range__date-inputs">
+			<legend className="date-range__date-inputs-legend">
+				{ translate( 'Start and End Dates' ) }
```
```suggestion
				{ translate( 'Start and end dates' ) }
```

Labels should be sentence-case, see conversation in p1676381941241629-slack-C9EJ7KSGH

kozer replied:
Should I update this component as well?
https://github.com/Automattic/wp-calypso/blob/trunk/client/components/date-range/inputs.tsx#L94

danielbachhuber replied:
@kozer Probably outside the scope of this PR. You could do so in a separate PR, though.

danielbachhuber commented on client/my-sites/hosting/web-server-logs-card/index.js:
```diff
@@ -60,54 +253,58 @@ const WebServerLogsCard = ( props ) => {
+			let startValidatorResults = {
+				isValid: startDateIsValid,
+				validationInfo: startDateIsValid
+					? translate( 'Date is valid' )
```
I don't think we need to display a message when the date is valid. It's a little bit redundant.

kozer replied:
I totally agree. And this was my initial approach.
However the `FormInputValidation` component, uses a checkmark in case of a non-error state, and it looks a bit ugly if there is no text there.
We can take another approach though, and hide the component in case we don't have an error.
If you think this is better, I can do that.

danielbachhuber replied:
> uses a checkmark in case of a non-error state, and it looks a bit ugly if there is no text there.
> We can take another approach though, and hide the component in case we don't have an error.
> If you think this is better, I can do that.

@kozer Ah, ok. I don't have a strong opinion about it and am fine with either option.

danielbachhuber reviewed (commented):
Sorry, a point of clarification: I should be able to pick _any_ 14 day date range.

It looks like I can't select dates more than 14 days ago?

<img width="843" alt="image" src="https://user-images.githubusercontent.com/36432/221850370-7d292e7b-67b8-4872-9c43-b6bafa16b95b.png">

sejas reviewed (commented):
Now it's working well, also in Spanish.

Another observation, I can't pick the same day for start and end dates. Is that expected?
As a user I would like to download the logs for "today".


<img width="617" alt="Screenshot 2023-02-28 at 16 42 53" src="https://user-images.githubusercontent.com/779993/221919994-34567a22-7573-4cd6-98e7-172cef227517.png">

kozer commented:
> Now it's working well, also in Spanish.
> 
> Another observation, I can't pick the same day for start and end dates. Is that expected? As a user I would like to download the logs for "today".
> 
> <img alt="Screenshot 2023-02-28 at 16 42 53" width="617" src="https://user-images.githubusercontent.com/779993/221919994-34567a22-7573-4cd6-98e7-172cef227517.png">

This was what we already had. I didn't change anything regarding the validations. If it's something we want, then ok, I can do that

danielbachhuber commented:
@kozer It might be helpful to write some tests for the validation behavior. Could be easier to keep track of how everything is expected to work.

sejas commented:
> This was what we already had. I didn't change anything regarding the validations. If it's something we want, then ok, I can do that

The issue comes from removing the hours. We probably need to set the start date to 00:00:00 and the end date to 23:59:59.

This is the behaviour we currently have in production. We can pick the same day but different hours:
<img width="970" alt="Screenshot 2023-03-01 at 14 45 10" src="https://user-images.githubusercontent.com/779993/222173764-0e09189f-355d-4d66-aa2f-016d82e87205.png">

sejas reviewed (approved):
It works as expected.

I've created this PR from your branch, where I use the start and end of the day so we can pick the same day in the date picker.

- https://github.com/Automattic/wp-calypso/pull/73945

@kozer  can you review it ?

sejas committed:
Developer UX: use start and end of the day to download logs (#73945)

kozer commented:
> It works as expected.
> 
> I've created this PR from your branch, where I use the start and end of the day so we can pick the same day in the data picker.
> 
> * #73945
> 
> @kozer can you review it ?

Thanks! I merged it in!

kozer committed:
Update: update start and end time when we ask for logs

kozer committed:
Update: update validation for start and end time when we ask for logs

danielbachhuber committed:
Sentence case for labels

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7837471

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7837471).
